finality-aleph/src/new_network/connection_manager/discovery.rs

```diff
@@ -0,0 +1,690 @@
+    pub fn new(cooldown: time::Duration) -> Self {
+        Discovery {
+            cooldown,
+            last_broadcast: HashMap::new(),
+            last_response: HashMap::new(),
+            requested_authentications: HashMap::new(),
+            next_query: NodeIndex(0),
+        }
+    }
+
+    /// Returns messages that should be sent as part of authority discovery at this moment.
+    pub fn discover_authorities(
+        &mut self,
+        handler: &SessionHandler,
+    ) -> Vec<(DiscoveryMessage, DataCommand)> {
+        let missing_authorities = handler.missing_nodes();
+        if missing_authorities.is_empty() {
+            return Vec::new();
+        }
+        let node_count = handler.node_count();
+        let authentication = handler.authentication();
+        if missing_authorities.len() * 3 > 2 * node_count.0 {
+            // We know of fewer than 1/3 authorities, broadcast our authentication and hope others
+            // respond in kind.
+            vec![authentication_broadcast(authentication)]
+        } else {
+            // Attempt learning about more authorities from the ones you already know.
+            let mut result = Vec::new();
+            let mut target = self.next_query;
+            while result.len() < 2 {
+                if let Some(peer_id) = handler.peer_id(&target) {
+                    result.push(request(
+                        missing_authorities.clone(),
+                        authentication.clone(),
+                        peer_id,
+                    ));
+                }
+                target = NodeIndex((target.0 + 1) % node_count.0);
+            }
+            self.next_query = target;
+            result
+        }
+    }
+
+    fn handle_authentication(
+        &mut self,
+        authentication: Authentication,
+        handler: &mut SessionHandler,
+    ) -> Vec<Multiaddr> {
+        if !handler.handle_authentication(authentication.clone()) {
+            return Vec::new();
+        }
+        authentication.0.address()
+    }
+
+    fn handle_broadcast(
+        &mut self,
+        authentication: Authentication,
+        handler: &mut SessionHandler,
+    ) -> (Vec<Multiaddr>, Option<(DiscoveryMessage, DataCommand)>) {
+        let addresses = self.handle_authentication(authentication.clone(), handler);
+        if addresses.is_empty() {
+            return (addresses, None);
+        }
+        let node_id = authentication.0.creator();
+        let rebroadcast = match self.last_broadcast.get(&node_id) {
+            Some(instant) => time::Instant::now() > *instant + self.cooldown,
+            None => true,
+        };
+        let message = match rebroadcast {
+            true => {
+                self.last_broadcast.insert(node_id, time::Instant::now());
+                Some(authentication_broadcast(authentication))
+            }
+            false => None,
+        };
+        (addresses, message)
```

Comment:
sorry, this is still a new context for me and I have some problems with connecting things and this is some sanity check: the node who gets a broadcast message not necessarily will interact with asked peers, i.e. `addresses`, so there may not be a need to keep connections with them, right?

Reply:
Yes, but this module does not contain logic that decides that – it always returns the addresses the node should be connected to if it wants to stay connected to the committee.

Reply:
Fair point, but then, to avoid such confusion, I would modify just a little bit the comment documenting `handle_message`, to emphasize that this: _should be connected to **if it wants to stay connected to the committee**_

Reply:
Added.